Test Server fails when starting a workflow that already exists.

I am running an integration test that is passing fine on both, a local docker-compose based server and on Temporal Cloud. But on the test server it fails immediately, and the server throws an internal exception.


* The test starts a workflow. (That succeeds.)

* Then it attempts to start a workflow with the _same_ Workflow Id as the workflow just started.

* The test expects to get an `AlreadyExists` gRPC status / error from the `StartWorkflowExecution(..)` gRPC call.

* Instead, it gets an unspecified gRPC error and the test server displays the following:
```
PS > .\temporal-test-server.exe 7233
May 14, 2022 2:06:22 AM io.grpc.internal.SerializingExecutor run
SEVERE: Exception while executing runnable io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed@6bb8cba3
java.lang.RuntimeException: Generated message class "io.temporal.api.errordetails.v1.WorkflowExecutionAlreadyStartedFailure" missing method "getStartRequestId".
        at com.google.protobuf.GeneratedMessageV3.getMethodOrDie(GeneratedMessageV3.java:1847)
        at com.google.protobuf.GeneratedMessageV3.access$1000(GeneratedMessageV3.java:79)
        at com.google.protobuf.GeneratedMessageV3$FieldAccessorTable$SingularFieldAccessor$ReflectionInvoker.<init>(GeneratedMessageV3.java:2176)
        at com.google.protobuf.GeneratedMessageV3$FieldAccessorTable$SingularFieldAccessor.<init>(GeneratedMessageV3.java:2248)
        at com.google.protobuf.GeneratedMessageV3$FieldAccessorTable$SingularStringFieldAccessor.<init>(GeneratedMessageV3.java:2916)
        at com.google.protobuf.GeneratedMessageV3$FieldAccessorTable.ensureFieldAccessorsInitialized(GeneratedMessageV3.java:1977)
        at io.temporal.api.errordetails.v1.WorkflowExecutionAlreadyStartedFailure.internalGetFieldAccessorTable(WorkflowExecutionAlreadyStartedFailure.java:93)
        at com.google.protobuf.GeneratedMessageV3.getDescriptorForType(GeneratedMessageV3.java:133)
        at com.google.protobuf.Any.pack(Any.java:183)
        at io.temporal.serviceclient.StatusUtils.newException(StatusUtils.java:82)
        at io.temporal.internal.testservice.TestWorkflowService.throwDuplicatedWorkflow(TestWorkflowService.java:286)
        at io.temporal.internal.testservice.TestWorkflowService.startWorkflowExecutionImpl(TestWorkflowService.java:231)
        at io.temporal.internal.testservice.TestWorkflowService.startWorkflowExecution(TestWorkflowService.java:204)
        at io.temporal.api.workflowservice.v1.WorkflowServiceGrpc$MethodHandlers.invoke(WorkflowServiceGrpc.java:3822)
        at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182)
        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:340)
        at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:866)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(Thread.java:833)
        at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:704)
        at com.oracle.svm.core.windows.WindowsPlatformThreads.osThreadStartRoutine(WindowsPlatformThreads.java:143)
Caused by: java.lang.NoSuchMethodException: io.temporal.api.errordetails.v1.WorkflowExecutionAlreadyStartedFailure.getStartRequestId()
        at java.lang.Class.getMethod(DynamicHub.java:2227)
        at com.google.protobuf.GeneratedMessageV3.getMethodOrDie(GeneratedMessageV3.java:1844)
        ... 23 more
```

--- a/temporal-serviceclient/src/main/java/io/temporal/serviceclient/StatusUtils.java
+++ b/temporal-serviceclient/src/main/java/io/temporal/serviceclient/StatusUtils.java
@@ -90,15 +90,8 @@ public class StatusUtils {
   private static <T extends GeneratedMessageV3> Any packAny(
       T details, Descriptors.Descriptor descriptor) {
     return Any.newBuilder()
-        .setTypeUrl(getTypeUrl("type.googleapis.com", descriptor))
+        .setTypeUrl("type.googleapis.com/" + descriptor.getFullName())
         .setValue(details.toByteString())
         .build();
-  }
-
-  private static String getTypeUrl(
-      java.lang.String typeUrlPrefix, com.google.protobuf.Descriptors.Descriptor descriptor) {
-    return typeUrlPrefix.endsWith("/")
-        ? typeUrlPrefix + descriptor.getFullName()
-        : typeUrlPrefix + "/" + descriptor.getFullName();
   }
 }

--- a/temporal-serviceclient/src/main/java/io/temporal/serviceclient/StatusUtils.java
+++ b/temporal-serviceclient/src/main/java/io/temporal/serviceclient/StatusUtils.java
@@ -21,9 +21,7 @@
 package io.temporal.serviceclient;
 
 import com.google.common.base.Preconditions;
-import com.google.protobuf.Any;
-import com.google.protobuf.GeneratedMessageV3;
-import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.*;
 import com.google.rpc.Status;
 import io.grpc.StatusRuntimeException;
 import io.grpc.protobuf.StatusProto;
@@ -74,14 +72,33 @@ public class StatusUtils {
 
   /** Create StatusRuntimeException with given details. */
   public static <T extends GeneratedMessageV3> StatusRuntimeException newException(
-      io.grpc.Status status, T details) {
+      io.grpc.Status status, T details, Descriptors.Descriptor detailsDescriptor) {
     Preconditions.checkNotNull(status, "status cannot be null");
     Status protoStatus =
         Status.newBuilder()
             .setCode(status.getCode().value())
             .setMessage(status.getDescription())
-            .addDetails(Any.pack(details))
+            .addDetails(packAny(details, detailsDescriptor))
             .build();
     return StatusProto.toStatusRuntimeException(protoStatus);
+  }
+
+  /**
+   * This method does exactly what {@link Any#pack(Message)} does. But it doesn't go into reflection
+   * to fetch the {@code descriptor}, which allows us to avoid a bunch of Graal reflection configs.
+   */
+  private static <T extends GeneratedMessageV3> Any packAny(
+      T details, Descriptors.Descriptor descriptor) {
+    return Any.newBuilder()
+        .setTypeUrl(getTypeUrl("type.googleapis.com", descriptor))
+        .setValue(details.toByteString())
+        .build();
+  }
+
+  private static String getTypeUrl(
+      java.lang.String typeUrlPrefix, com.google.protobuf.Descriptors.Descriptor descriptor) {
+    return typeUrlPrefix.endsWith("/")
+        ? typeUrlPrefix + descriptor.getFullName()
+        : typeUrlPrefix + "/" + descriptor.getFullName();
   }
 }

--- a/temporal-test-server/src/main/java/io/temporal/internal/testservice/StateMachines.java
+++ b/temporal-test-server/src/main/java/io/temporal/internal/testservice/StateMachines.java
@@ -1398,7 +1398,8 @@ class StateMachines {
             result.completeExceptionally(
                 StatusUtils.newException(
                     Status.INTERNAL.withDescription(value.getErrorMessage()),
-                    QueryFailedFailure.getDefaultInstance()));
+                    QueryFailedFailure.getDefaultInstance(),
+                    QueryFailedFailure.getDescriptor()));
             break;
           default:
             throw Status.INVALID_ARGUMENT

--- a/temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java
+++ b/temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java
@@ -536,7 +536,8 @@ class TestWorkflowMutableStateImpl implements TestWorkflowMutableState {
                         .completeExceptionally(
                             StatusUtils.newException(
                                 Status.INTERNAL.withDescription(result.getErrorMessage()),
-                                QueryFailedFailure.getDefaultInstance()));
+                                QueryFailedFailure.getDefaultInstance(),
+                                QueryFailedFailure.getDescriptor()));
                     break;
                   case UNRECOGNIZED:
                     throw Status.INVALID_ARGUMENT
@@ -2171,7 +2172,8 @@ class TestWorkflowMutableStateImpl implements TestWorkflowMutableState {
         StatusRuntimeException error =
             StatusUtils.newException(
                 Status.INVALID_ARGUMENT.withDescription(completeRequest.getErrorMessage()),
-                QueryFailedFailure.getDefaultInstance());
+                QueryFailedFailure.getDefaultInstance(),
+                QueryFailedFailure.getDescriptor());
         result.completeExceptionally(error);
         break;
     }

--- a/temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java
+++ b/temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java
@@ -286,7 +286,8 @@ public final class TestWorkflowService extends WorkflowServiceGrpc.WorkflowServi
         Status.ALREADY_EXISTS.withDescription(
             String.format(
                 "WorkflowId: %s, " + "RunId: %s", execution.getWorkflowId(), execution.getRunId())),
-        error);
+        error,
+        WorkflowExecutionAlreadyStartedFailure.getDescriptor());
   }
 
   private StartWorkflowExecutionResponse startWorkflowExecutionNoRunningCheckLocked(

